File: sdk/typescript/src/flow.ts
/**
 * FlowProducer: atomic parent/children trees, chains and fan-in flows.
 */

import { Connection } from './connection.js';
import { CommandError } from './errors.js';
import { commitFlow } from './flow-commit.js';
import { type PlannedFlowNode, planFlows } from './flow-plan.js';
import { planBulkThen, planChain } from './flow-plan-legacy.js';
import type {
  FlowJob,
  FlowOptions,
  FlowProducerOptions,
  FlowStep,
  GetFlowOptions,
  JobNode,
} from './flow-types.js';
import { Job } from './job.js';
import type { JobResponse } from './responses.js';

export class FlowProducer {
  readonly connection: Connection;
  private readonly ownsConnection: boolean;

  constructor(opts: FlowProducerOptions = {}) {
    this.connection =
      opts.connection ??
      new Connection({
        host: opts.host,
        port: opts.port,
        token: opts.token,
        tls: opts.tls,
        logger: opts.logger,
        onTelemetry: opts.onTelemetry,
      });
    this.ownsConnection = opts.connection === undefined;
  }

  /** Add a flow tree in one broker-side atomic commit. */
  async add<T = unknown>(flow: FlowJob<T>, options?: FlowOptions): Promise<JobNode<T>> {
    const plan = planFlows([flow], options);
    const snapshots = await commitFlow(this.connection, plan.jobs);
    return this.buildNode(plan.roots[0], snapshots);
  }

  async addBulk<T = unknown>(flows: FlowJob<T>[]): Promise<JobNode<T>[]> {
    const plan = planFlows(flows);
    const snapshots = await commitFlow(this.connection, plan.jobs);
    return plan.roots.map((root) => this.buildNode(root, snapshots));
  }

  /** Fetch a flow tree starting from a job id (recursive over childrenIds). */
  getFlow<T = unknown>(opts: GetFlowOptions): Promise<JobNode<T> | null> {
    return this.fetchNode<T>(
      opts.id,
      opts.depth ?? Number.POSITIVE_INFINITY,
      opts.maxChildren,
      new Set()
    );
  }

  /** Add a sequential chain: step[0] → step[1] → ... via dependsOn. */
  async addChain<T = unknown>(steps: FlowStep<T>[]): Promise<{ jobIds: string[] }> {
    if (steps.length === 0) return { jobIds: [] };
    const plan = planChain(steps);
    await commitFlow(this.connection, plan.jobs);
    return { jobIds: plan.ids };
  }

  /** Parallel jobs converging into a final fan-in job. */
  async addBulkThen<T = unknown>(
    parallel: FlowStep<T>[],
    final: FlowStep<T>
  ): Promise<{ parallelIds: string[]; finalId: string }> {
    const plan = planBulkThen(parallel, final);
    await commitFlow(this.connection, plan.jobs);
    return { parallelIds: plan.parallelIds, finalId: plan.finalId };
  }

  close(): void {
    if (this.ownsConnection) this.connection.close();
  }

  // ----------------------------------------------------------------- internals

  private buildNode<T>(
    node: PlannedFlowNode<T>,
    snapshots: ReadonlyMap<string, Record<string, unknown>>
  ): JobNode<T> {
    const snapshot = snapshots.get(node.id);
    if (!snapshot) throw new Error(`Committed flow snapshot missing for ${node.id}`);
    const children = node.children?.map((child) => this.buildNode(child, snapshots));
    return {
      job: new Job<T>(snapshot, this.connection),
      children: children && children.length > 0 ? children : undefined,
    };
  }

  private async fetchNode<T>(
    id: string,
    depth: number,
    maxChildren: number | undefined,
    visited: Set<string>
  ): Promise<JobNode<T> | null> {
    if (visited.has(id)) return null; // cycle guard: id already on the current path
    visited.add(id);
    // A missing job — the root, or a child removed via removeOnComplete/cancel
    // (childrenIds is a static push-time list, never pruned) — yields null and
    // is skipped, returning the surviving partial tree instead of throwing.
    let response: JobResponse;
    try {
      response = await this.connection.call<JobResponse>({ cmd: 'GetJob', id });
    } catch (err) {
      // Only 'Job not found' means a removed node; a real server error must not
      // masquerade as a missing child and yield a misleading partial tree.
      if (err instanceof CommandError && /not found/i.test(err.message)) return null;
      throw err;
    }
    const raw = response.job;
    if (!raw) return null;
    const job = new Job<T>(raw, this.connection);
    if (depth <= 0 || job.childrenIds.length === 0) return { job };

    const limit = maxChildren ?? job.childrenIds.length;
    const children: JobNode<T>[] = [];
    for (const childId of job.childrenIds.slice(0, limit)) {
      const child = await this.fetchNode<T>(childId, depth - 1, maxChildren, visited);
      if (child) children.push(child);
    }
    return { job, children: children.length > 0 ? children : undefined };
  }
}

File: sdk/python/bunqueue/flow.py
"""FlowProducer: atomic trees, chains, fan-in flows, and flow reads."""

from __future__ import annotations

from typing import Any, Dict, List, Optional, Sequence, Set

from .connection import Connection, TlsOption
from .errors import CommandError
from .flow_commit import commit_flow
from .flow_plan import PlannedFlowNode, plan_flows
from .flow_plan_legacy import plan_bulk_then, plan_chain
from .job import Job
from .telemetry import TelemetryHandler


class FlowNode:
    """Result node built from the broker's committed job snapshot."""

    def __init__(self, job: Job, children: Optional[List["FlowNode"]] = None) -> None:
        self.job = job
        self.children = children or []


class FlowProducer:
    """Create dependent job graphs through one broker-side atomic commit."""

    def __init__(
        self,
        host: str = "localhost",
        port: int = 6789,
        token: Optional[str] = None,
        tls: TlsOption = None,
        connection: Optional[Connection] = None,
        on_telemetry: Optional[TelemetryHandler] = None,
    ) -> None:
        self.connection = connection or Connection(
            host=host, port=port, token=token, tls=tls, on_telemetry=on_telemetry
        )
        self._owns_connection = connection is None

    def add(
        self, flow: Dict[str, Any], opts: Optional[Dict[str, Any]] = None
    ) -> FlowNode:
        """Add a tree atomically; children become runnable before their parent."""
        plan = plan_flows([flow], opts)
        snapshots = commit_flow(self.connection, plan.jobs)
        return self._build_node(plan.roots[0], snapshots)

    def add_bulk(self, flows: Sequence[Dict[str, Any]]) -> List[FlowNode]:
        """Add multiple independent trees in one atomic batch."""
        plan = plan_flows(list(flows))
        snapshots = commit_flow(self.connection, plan.jobs)
        return [self._build_node(root, snapshots) for root in plan.roots]

    def add_chain(self, steps: Sequence[Dict[str, Any]]) -> List[str]:
        """Add a sequential dependency chain in one atomic batch."""
        if not steps:
            return []
        plan = plan_chain(steps)
        commit_flow(self.connection, plan.jobs)
        return plan.ids

    def add_bulk_then(
        self, parallel: Sequence[Dict[str, Any]], final: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Add parallel jobs and their final fan-in job in one atomic batch."""
        plan = plan_bulk_then(parallel, final)
        commit_flow(self.connection, plan.jobs)
        return {"parallel_ids": plan.parallel_ids, "final_id": plan.final_id}

    def get_flow(self, job_id: str, depth: Optional[int] = None) -> Optional[FlowNode]:
        """Reconstruct a surviving flow tree from broker snapshots."""
        return self._get_flow(job_id, depth, set())

    def close(self) -> None:
        if self._owns_connection:
            self.connection.close()

    def __enter__(self) -> "FlowProducer":
        return self

    def __exit__(self, *exc: Any) -> None:
        self.close()

    def _build_node(
        self,
        node: PlannedFlowNode,
        snapshots: Dict[str, Dict[str, Any]],
    ) -> FlowNode:
        snapshot = snapshots.get(node.id)
        if snapshot is None:
            raise ValueError(f"Committed flow snapshot missing for {node.id}")
        children = [self._build_node(child, snapshots) for child in node.children]
        return FlowNode(Job(snapshot, self.connection), children)

    def _get_flow(
        self, job_id: str, depth: Optional[int], visited: Set[str]
    ) -> Optional[FlowNode]:
        if job_id in visited:
            return None
        visited.add(job_id)
        try:
            response = self.connection.call({"cmd": "GetJob", "id": job_id})
        except CommandError as exc:
            if "not found" in str(exc).lower():
                return None
            raise
        raw = response.get("job")
        if not raw:
            return None
        job = Job(raw, self.connection)
        children: List[FlowNode] = []
        if depth is None or depth > 0:
            next_depth = None if depth is None else depth - 1
            for child_id in job.children_ids:
                child = self._get_flow(child_id, next_depth, visited)
                if child:
                    children.append(child)
        return FlowNode(job, children)
